src/Microsoft.ML.SearchSpace/Parameter.cs

```diff
@@ -0,0 +1,278 @@
+            Converters = new JsonConverter[]
+            {
+                new StringEnumConverter(),
+            },
+        };
+
+        private readonly object _value;
+
+        private Parameter(object value, ParameterType type)
+        {
+            this._value = value;
+            this.ParameterType = type;
+        }
+
+        public static Parameter FromDouble(double value)
+        {
+            return new Parameter(value, ParameterType.Float);
+        }
+
+        public static Parameter FromFloat(float value)
+        {
+            return new Parameter(value, ParameterType.Float);
+        }
+
+        public static Parameter FromLong(long value)
+        {
+            return new Parameter(value, ParameterType.Integer);
+        }
+
+        public static Parameter FromInt(int value)
+        {
+            return new Parameter(value, ParameterType.Integer);
+        }
+
+        public static Parameter FromString(string value)
+        {
+            return new Parameter(value, ParameterType.Integer);
+        }
+
+        public static Parameter FromBool(bool value)
+        {
+            return new Parameter(value, ParameterType.Bool);
+        }
+
+        public static Parameter FromEnum<T>(T value) where T : struct, Enum
+        {
+            return Parameter.FromEnum(value, typeof(T));
+        }
+
+        public static Parameter FromIEnumerable<T>(IEnumerable<T> values)
+        {
+            // check T
+            return Parameter.FromIEnumerable(values as IEnumerable);
+        }
+
+        private static Parameter FromIEnumerable(IEnumerable values)
+        {
+            return new Parameter(values, ParameterType.Array);
+        }
+
+        private static Parameter FromEnum(Enum e, Type t)
+        {
+            return Parameter.FromString(Enum.GetName(t, e));
+        }
+
+        public static Parameter FromObject<T>(T value) where T : class
+        {
+            return Parameter.FromObject(value, typeof(T));
+        }
+
+        private static Parameter FromObject(object value, Type type)
+        {
+            var param = value switch
+            {
+                int i => Parameter.FromInt(i),
+                long l => Parameter.FromLong(l),
+                double d => Parameter.FromDouble(d),
+                float f => Parameter.FromFloat(f),
+                string s => Parameter.FromString(s),
+                bool b => Parameter.FromBool(b),
+                IEnumerable vs => Parameter.FromIEnumerable(vs),
+                Enum e => Parameter.FromEnum(e, e.GetType()),
+                _ => null,
+            };
+
+            if (param != null)
+            {
+                return param;
+            }
+            else
+            {
+                var parameter = Parameter.CreateNestedParameter();
+                var properties = type.GetProperties(System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance)
```

Comment:
You need to add some checks here for unsupported types. I could pass in something like `uint`, which isn't caught by the switch up top, and I believe it would cause issues. (Though if you have testing showing it works fine than nothing is needed to be added.

Reply:
You can do that as long as the value can be converted to `int`. It might be confusing though but underlying it's using Json.Net to map a `IParameter` to a specific type so as long as the value can be casted, or that specific type implements `ISerializable` or has `JsonConverter`, the converion will work

Reply:
### Resolved

Reply:
So are there any types this wouldn't work for? It seems like there would still be some?

Reply:
Yeah there would be, for example the cast can't be accomplished if you try to cast a string to an enum which string value isn't one of them

Reply:
So is that something you need to have a check for? just in case it can't be done?

Reply:
There's no easier way to check for that error since it's not obvious. We need to go through that enum and compare string value one by one, in which case I feel it's no better than just throw an exception when an casting error get caught.

I add check for obvious errors though, like
- use `UniformSingleOption` or `UniformDoubleOption` for integar type.
- apply `BooleanAttribute` to non-bool property type.
- apply `ChoiceAttribute` to unsupported type.

Reply:
### Resolved